### PR TITLE
machine/atsamd21: support for USB CDC aka serial interface

### PR DESCRIPTION
**Pull request by deadprogram**
This PR adds support for the USB CDC aka serial port over USB interface.

### CODE REVIEW

File: src/machine/machine_atsamd21.go [hunk 1]
```diff
@@ -102,11 +118,15 @@ type UART struct {
 }
 
 var (
+	// UART0 is actually a USB CDC interface.
+	//UART0 = UART{Buffer: NewRingBuffer(), UsesUSB: true}
```
Comment (aykevl) [inline src/machine/machine_atsamd21.go]: Is this comment here intentional, with `UsesUSB`?

Comment (deadprogram) [reply src/machine/machine_atsamd21.go]: No that was a comment that should have been removed.

File: src/machine/machine_atsamd21.go [hunk 2]
```diff
@@ -906,3 +926,909 @@ func (pwm PWM) setChannel(val sam.RegValue) {
...
+
+	usb_DEVICE_PCKSIZE_SIZE_Pos  = 28
+	usb_DEVICE_PCKSIZE_SIZE_Mask = 0x7
+
+	usb_DEVICE_PCKSIZE_MULTI_PACKET_SIZE_Pos  = 14
+	usb_DEVICE_PCKSIZE_MULTI_PACKET_SIZE_Mask = 0x3FFF
+)
+
+var (
+	//go:volatile
```
Comment (aykevl) [inline src/machine/machine_atsamd21.go]: This... works? I don't remember writing code that marks these as volatile.

File: src/machine/machine_atsamd21.go [hunk 3]
```diff
@@ -906,3 +926,909 @@ func (pwm PWM) setChannel(val sam.RegValue) {
...
+	sam.USB_DEVICE.CTRLA |= sam.USB_DEVICE_CTRLA_SWRST
+	for (sam.USB_DEVICE.SYNCBUSY&sam.USB_DEVICE_SYNCBUSY_SWRST) > 0 ||
+		(sam.USB_DEVICE.SYNCBUSY&sam.USB_DEVICE_SYNCBUSY_ENABLE) > 0 {
+	}
+
+	sam.USB_DEVICE.DESCADD = sam.RegValue(uintptr(unsafe.Pointer(&usbEndpointDescriptors)))
+
+	// configure pins
+	GPIO{24}.Configure(GPIOConfig{Mode: GPIO_COM})
+	GPIO{25}.Configure(GPIOConfig{Mode: GPIO_COM})
```
Comment (aykevl) [inline src/machine/machine_atsamd21.go]: Are these pins used on all SAMD21 chips, on all boards?

File: src/machine/machine_atsamd21.go [hunk 4]
```diff
@@ -906,3 +926,909 @@ func (pwm PWM) setChannel(val sam.RegValue) {
...
+	case 64:
+		return 3
+	case 128:
+		return 4
+	case 256:
+		return 5
+	case 512:
+		return 6
+	case 1023:
+		return 7
```
Comment (aykevl) [inline src/machine/machine_atsamd21.go]: 1023? Not 1024?

Comment (deadprogram) [reply src/machine/machine_atsamd21.go]: Yes, that is on the data sheet exactly that way.

File: src/machine/machine_atsamd21.go (hunk 3 above)
Comment (deadprogram) [reply src/machine/machine_atsamd21.go]: No, that should be a board constant.

File: src/machine/machine_atsamd21.go (hunk 4 above)
Comment (aykevl) [reply src/machine/machine_atsamd21.go]: Well, if they say so...

File: src/machine/machine_atsamd21.go (hunk 2 above)
Comment (deadprogram) [reply src/machine/machine_atsamd21.go]: I think it just does nothing. Removing.

**Comment by deadprogram** [comment]
Made requested corrections and some small refactors for API consistency.

**Comment by aykevl** [comment]
Merged in 5438f16fcb1b541715b0717d79d78fb75123a18a.

I think the GC is used more than necessary at the moment, but I don't see an obvious solution so I'll leave that until a later time, perhaps when allocation elision gets better or it is necessary to fix the code itself.

**Pull request closed without merging**